### PR TITLE
Feature/1-interaction-protocol

### PR DESCRIPTION
**Pull request by PopeyeTheSailorsCat**
I think I need to show my work already at this stage. Since I seem to have implemented what was intended, but whether I understood it correctly and whether I implemented it correctly, I do not know.

### CODE REVIEW

File: gRPC/spec.proto [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+
+service Modeling {
+    rpc GetUpdate (UpdateRequest) returns (stream UpdateResponse) {}
+    rpc GetMap (Empty) returns (stream Map) {}
+}
+
+message Metadata {
+    enum statusCode{
+        SUCCESS = 0;
```
Comment (yarnaid) [inline gRPC/spec.proto]: The first value will be the default. Let's use this UNDEFINED in addition to others

File: gRPC/spec.proto [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        ERROR = 1;
+    }
+    statusCode status = 1;
+    int32  request_id = 2;
```
Comment (yarnaid) [inline gRPC/spec.proto:16]: Let's use UUID for this. In our case, will be enough just a string without any validation
Also, let's put `request_id` to the Meta message

File: gRPC/spec.proto [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    Metadata meta = 1;
+    State state = 2;
+
+}
+
+message Map {
+    Metadata meta = 1;
+    // get buildings
+    message Building{
+        int32 id=1; //unique key
```
Comment (yarnaid) [inline gRPC/spec.proto]: These values are duplicated from the `State` massage. Let's make a bit more complex structure.
We can use something like `BaseUnit` to store universal values for each entity and then extend them

File: gRPC/spec.proto [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        int32 type = 2; // building specification (house, road, clinic)
```
Comment (yarnaid) [inline gRPC/spec.proto]: why the type is int?

File: tests/client_server.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+import unittest
```
Comment (yarnaid) [inline tests/client_server.py]: let's use `pytest`

File: backend/server.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+class ModelingServicer(spec_pb2_grpc.ModelingServicer):
+
+    def __init__(self, map_obj, model_objects):  # Remember our Backend output exists
+        self.map = map_obj
+        self.model_objects = model_objects
+
+    def GetUpdate(self, request, context):  # Generator of people on modeling
+        status = spec_pb2.Metadata(status=gRPC_status_creator["SUCCESS"],
+                                   request_id=request.meta.request_id)
```
Comment (yarnaid) [inline backend/server.py]: let's use [this](https://pypi.org/project/pure-protobuf/) to simplify massages processing

File: backend/server.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    @staticmethod
+    def create_update_response(human, status):  # create pb2 object from backed human
+        return spec_pb2.UpdateResponse(meta=status,
+                                       state=spec_pb2.State(
+                                           id=human.id,
+                                           type=human.type,
+                                           coord_x=human.x,
+                                           coord_y=human.y
+                                       ))
```
Comment (yarnaid) [inline backend/server.py]: Let's follow [MVC](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller). Models are our messages, views are a request-response server and controllers we must extract from the views leyer

File: backend/server.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+                                coord_x=building.x,
+                                coord_y=building.y,
+                                width=building.width,
+                                length=building.length,
+                                angle=building.angle
+                            ))
+
+
+def serve():  # Responsible for the operation of the server
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
```
Comment (yarnaid) [inline backend/server.py]: [cpu count](https://devdocs.io/python~3.8/library/os#os.cpu_count)

File: backend/server.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    spec_pb2_grpc.add_ModelingServicer_to_server(
+        ModelingServicer(None, None), server)  # NEED BACKEND OUTPUT INSTEAD OF NONE
+    server.add_insecure_port('[::]:50051')
+    server.start()
+    server.wait_for_termination()
+
+
+if __name__ == '__main__':
+    logging.basicConfig()
```
Comment (yarnaid) [inline backend/server.py]: Let's create a dedicated logger for each file. It makes debugging simpler. Also, usage of logging module is not the best practice

File: frontend/client.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
+from __future__ import print_function
```
Comment (yarnaid) [inline frontend/client.py]: let's rename this to `echo_client` and add debug output. For FE we will have another client)

**Review by yarnaid — CHANGES_REQUESTED** [review]

**Comment by yarnaid** [comment]
Great start, thank you! Could you also add a `Makefile` to simplify and unify the run? Please, add there commands for grpc code generation, running test with/without coverage

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,7 @@
+create-proto:
+	python -m grpc_tools.protoc -IgRPC --python_out=gRPC --grpc_python_out=gRPC gRPC/spec.proto
+test-coverage:
+	python -m pytest --cov=backend --cov=frontend tests/
+
+test-no-coverage:
```
Comment (yarnaid) [inline Makefile]: let's call this just `test`

File: dataStructure/gRPC.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+# gRPC_building_types = {'HOUSE': 0, 'ROAD': 1}
+
+
+class RequestCounter:  # counter of unique request id
+    id = 0
+
+    @staticmethod
+    def give_id():
+        RequestCounter.id += 1
+        return RequestCounter.id
```
Comment (yarnaid) [inline dataStructure/gRPC.py]: let's use [this](https://devdocs.io/python~3.8/library/itertools#itertools.count)

File: dataStructure/gRPC.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    id: int32 = field(1, default=int32(0))
+    coord_x: int32 = field(2, default=int32(0))
+    coord_y: int32 = field(3, default=int32(0))
+
+
+class HumanType(IntEnum):
+    NORMAL = 0
+    ILL = 1
+    RECOVERED = 2
+    DEAD = 3
```
Comment (yarnaid) [inline dataStructure/gRPC.py:34]: 👍

File: dataStructure/gRPC.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+
+@message
+@dataclass
+class State:
```
Comment (yarnaid) [inline dataStructure/gRPC.py]: class name `State` can be a bit confusing. Could you make it more exact

File: dataStructure/gRPC.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    width: int32 = field(3)
+    length: int32 = field(4)
+    angle: int32 = field(5)
+
+
+@message
+@dataclass
+class Map:
+    meta: Metadata = field(1)
+    building: Building = field(2)
```
Comment (yarnaid) [inline dataStructure/gRPC.py]: I think that it should be an array

**Review by yarnaid — CHANGES_REQUESTED** [review]

File: dataStructure/gRPC.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
+from dataclasses import dataclass
+from pure_protobuf.dataclasses_ import field, message
+from pure_protobuf.types import int32
+from enum import IntEnum
+
+
+# gRPC_status_creator = {'UNDEFINED': 0, 'SUCCESS': 1, 'ERROR': 2}  # Dont know how to do it better
+# gRPC_human_types = {'NORMAL': 0, 'ILL': 1, 'RECOVERED': 2, 'DEAD': 3}
+# gRPC_building_types = {'HOUSE': 0, 'ROAD': 1}
```
Comment (yarnaid) [inline dataStructure/gRPC.py]: let's remove comments

File: gRPC/spec.proto (hunk 2 above)
Comment (PopeyeTheSailorsCat) [reply gRPC/spec.proto:16]: Done

File: backend/server.py (hunk 7 above)
Comment (PopeyeTheSailorsCat) [reply backend/server.py]: Done(excluding echo_client)

File: backend/server.py (hunk 9 above)
Comment (PopeyeTheSailorsCat) [reply backend/server.py]: Done?

File: Makefile (hunk 11 above)
Comment (PopeyeTheSailorsCat) [reply Makefile]: Done

File: dataStructure/gRPC.py (hunk 12 above)
Comment (PopeyeTheSailorsCat) [reply dataStructure/gRPC.py]: Done

File: dataStructure/gRPC.py (hunk 14 above)
Comment (PopeyeTheSailorsCat) [reply dataStructure/gRPC.py]: Done

File: dataStructure/gRPC.py (hunk 16 above)
Comment (PopeyeTheSailorsCat) [reply dataStructure/gRPC.py]: Done

File: dataStructure/gRPC.py (hunk 15 above)
Comment (PopeyeTheSailorsCat) [reply dataStructure/gRPC.py]: Done

File: backend/server.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+import logging
+import grpc
+from dataStructure.gRPC import gRPC_status_creator
+
+
+class ModelingServicer(spec_pb2_grpc.ModelingServicer):
+
+    def __init__(self, map_obj, model_objects):  # Remember our Backend output exists
+        self.map = map_obj
+        self.model_objects = model_objects
```
Comment (yarnaid) [inline backend/server.py]: could you rename this variable to be more clear

File: frontend/client.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+from gRPC import spec_pb2_grpc, spec_pb2
+from dataStructure.gRPC import gRPC_status_creator, RequestCounter
+import logging
+
+import grpc
+
+
+def update_request(stub):  # Request to update the states of modeling objects
```
Comment (yarnaid) [inline frontend/client.py]: please, replace these comments with docstrings

**Review by yarnaid — CHANGES_REQUESTED** [review]

File: backend/server.py (hunk 17 above)
Comment (PopeyeTheSailorsCat) [reply backend/server.py]: Done

File: frontend/client.py (hunk 18 above)
Comment (PopeyeTheSailorsCat) [reply frontend/client.py]: Done?

**Pull request merged**